rllib/agents/trainer.py

```diff
@@ -927,16 +982,105 @@ def step(self) -> ResultDict:
                 # Allow logs messages to propagate.
                 time.sleep(0.5)
                 raise e
+
+            # Stopping criteria.
+            if self.config["_disable_distributed_execution_api"]:
+                if self._by_agent_steps:
+                    sampled = self._counters[NUM_AGENT_STEPS_SAMPLED] - \
+                        agent_steps_sampled
+                    trained = self._counters[NUM_AGENT_STEPS_TRAINED] - \
+                        agent_steps_trained
+                else:
+                    sampled = self._counters[NUM_ENV_STEPS_SAMPLED] - \
+                        env_steps_sampled
+                    trained = self._counters[NUM_ENV_STEPS_TRAINED] - \
+                        env_steps_trained
+
+                min_t = self.config["min_time_s_per_reporting"]
+                min_sample_ts = self.config[
+                    "min_sample_timesteps_per_reporting"]
+                min_train_ts = self.config["min_train_timesteps_per_reporting"]
+                # Repeat if not enough time has passed or if not enough
+                # env|train timesteps have been processed (or these min values
+                # are not provided by the user).
+                if (not min_t or time.time() - time_start >= min_t) and \
+                        (not min_sample_ts or sampled >= min_sample_ts) and \
+                        (not min_train_ts or trained >= min_train_ts):
+                    break
             else:
                 break
 
-        # Still no result (even after n retries).
-        if result is None:
-            raise RuntimeError("Failed to recover from worker crash.")
-
         if hasattr(self, "workers") and isinstance(self.workers, WorkerSet):
+            # Sync filters on workers.
             self._sync_filters_if_needed(self.workers)
 
+            # Collect worker metrics.
+            if self.config["_disable_distributed_execution_api"]:
+                info = result.copy()
+                result["info"] = {LEARNER_INFO: info}
```

Comment:
hmm, we recursively save another copy of result under info key??

Reply:
We do. I think I did this for backward compatibility. ... :/ Yeah, the `result`structure itself needs its own cleanup PR, I think. We should talk this through though before we start making changes to this. Many users could relay on certain keys being where they expect them to be.